akka-javasdk/src/main/scala/akka/javasdk/impl/workflow/ReflectiveWorkflowRouter.scala

```diff
@@ -4,63 +4,199 @@
 
 package akka.javasdk.impl.workflow
 
+import java.util.concurrent.CompletionStage
+import java.util.function.{ Function => JFunc }
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.jdk.FutureConverters.CompletionStageOps
+import scala.jdk.OptionConverters.RichOptional
+
 import akka.annotation.InternalApi
 import akka.javasdk.impl.AnySupport
 import akka.javasdk.impl.CommandHandler
 import akka.javasdk.impl.CommandSerialization
 import akka.javasdk.impl.InvocationContext
+import akka.javasdk.impl.WorkflowExceptions.WorkflowException
+import akka.javasdk.impl.serialization.JsonSerializer
+import akka.javasdk.impl.workflow.ReflectiveWorkflowRouter.CommandHandlerNotFound
+import akka.javasdk.impl.workflow.ReflectiveWorkflowRouter.CommandResult
+import akka.javasdk.impl.workflow.ReflectiveWorkflowRouter.WorkflowStepNotFound
+import akka.javasdk.impl.workflow.ReflectiveWorkflowRouter.WorkflowStepNotSupported
+import akka.javasdk.timer.TimerScheduler
 import akka.javasdk.workflow.CommandContext
 import akka.javasdk.workflow.Workflow
-import com.google.protobuf.any.{ Any => ScalaPbAny }
+import akka.javasdk.workflow.Workflow.AsyncCallStep
+import akka.javasdk.workflow.Workflow.Effect
+import akka.runtime.sdk.spi.BytesPayload
+import akka.runtime.sdk.spi.SpiWorkflow
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+object ReflectiveWorkflowRouter {
+  final case class CommandResult(effect: Workflow.Effect[_])
+
+  final case class CommandHandlerNotFound(commandName: String) extends RuntimeException {
+    override def getMessage: String = commandName
+  }
+  final case class WorkflowStepNotFound(stepName: String) extends RuntimeException {
+    override def getMessage: String = stepName
+  }
+
+  final case class WorkflowStepNotSupported(stepName: String) extends RuntimeException {
+    override def getMessage: String = stepName
+  }
+}
 
 /**
  * INTERNAL API
  */
 @InternalApi
 class ReflectiveWorkflowRouter[S, W <: Workflow[S]](
-    override protected val workflow: W,
-    commandHandlers: Map[String, CommandHandler])
-    extends WorkflowRouter[S, W](workflow) {
+    val workflow: W,
+    commandHandlers: Map[String, CommandHandler],
+    serializer: JsonSerializer) {
+
+  private def decodeUserState(userState: Option[BytesPayload]): S =
+    userState
+      .collect {
+        case payload if payload != BytesPayload.empty => serializer.fromBytes(payload).asInstanceOf[S]
+      }
+      // if runtime doesn't have a state to provide, we fall back to user's own defined empty state
+      .getOrElse(workflow.emptyState())
```

Comment:
@patriknw, comment here for the context. 

Here is how it was and how it is now. 
The SPI is not providing a `emptyState` method anymore. It never used it. 

The runtime passes a Some or a None for the state. When it's a None, the SDK fallback to user's defined `emptyState` (which can be null). As seen here.

The decode method is not protect with `payload != BytesPayload.empty`, but in principle it can never be `BytesPayload.empty` because `Option[BytesPayload]` always come from runtime.